File: libs/shape-clipboard/src/main/kotlin/mono/shape/clipboard/ShapeClipboardManager.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.shape.clipboard

import kotlinx.browser.document
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import mono.graphics.geo.Rect
import mono.html.TextArea
import mono.livedata.LiveData
import mono.livedata.MutableLiveData
import mono.shape.extra.TextExtra
import mono.shape.serialization.AbstractSerializableShape
import mono.shape.serialization.SerializableLineConnector
import mono.shape.serialization.SerializableText
import org.w3c.dom.HTMLElement

/**
 * A clipboard manager specializing for shapes.
 * This class handles storing shapes to clipboard and getting shapes in clipboard from paste action.
 */
class ShapeClipboardManager(private val body: HTMLElement) {

    private val clipboardShapeMutableLiveData: MutableLiveData<List<AbstractSerializableShape>> =
        MutableLiveData(emptyList())
    val clipboardShapeLiveData: LiveData<List<AbstractSerializableShape>> =
        clipboardShapeMutableLiveData

    init {
        document.onpaste = {
            it.preventDefault()
            it.stopPropagation()
            onPasteText(it.clipboardData?.getData("text/plain").orEmpty())
        }
    }

    private fun onPasteText(text: String) {
        if (text.isBlank()) {
            return
        }
        val clipboardObject = try {
            Json.decodeFromString<ClipboardObject>(text)
        } catch (e: Exception) {
            null
        }
        if (clipboardObject != null) {
            clipboardShapeMutableLiveData.value = clipboardObject.shapes
            // TODO: Handle connectors
            return
        }

        // This is for backward compatibility with old clipboard format
        val clipboardShapes = try {
            Json.decodeFromString<List<AbstractSerializableShape>>(text)
        } catch (e: Exception) {
            null
        }
        if (clipboardShapes != null) {
            clipboardShapeMutableLiveData.value = clipboardShapes
            return
        }

        clipboardShapeMutableLiveData.value = listOf(createTextShapeFromText(text))
    }

    private fun createTextShapeFromText(text: String): SerializableText {
        val lines = text
            .split('\n')
            .flatMap { it.chunked(DEFAULT_TEXT_BOUND_WIDTH) }
        val width = lines.maxOf { it.length }
        val height = lines.size

        // Replace space chars with nbsp chars to avoid space chars are being trimmed by the browser
        val toBeUsedText = text.replace(' ', NON_BREAKING_SPACE_CHAR)
        return SerializableText(
            bound = Rect.Companion.byLTWH(0, 0, width, height),
            text = toBeUsedText,
            extra = TextExtra.NO_BOUND.toSerializableExtra()
        )
    }

    fun setClipboard(
        shapes: List<AbstractSerializableShape>,
        connectors: List<SerializableLineConnector>
    ) {
        val clipboardObject = ClipboardObject(shapes, connectors)
        val json = Json.encodeToString(clipboardObject)
        setClipboardText(json)
    }

    fun setClipboardText(text: String) {
        TextArea(body, classes = "hidden", content = text) {
            select()
            document.execCommand("copy")
            remove()
        }
    }

    /**
     * A data class to store shapes and connectors in clipboard when serializing to JSON.
     */
    @Serializable
    class ClipboardObject(
        val shapes: List<AbstractSerializableShape>,
        val connectors: List<SerializableLineConnector>
    )

    companion object {
        private const val DEFAULT_TEXT_BOUND_WIDTH = 400
        private const val NON_BREAKING_SPACE_CHAR = '\u00a0'
    }
}

File: monosketch-svelte/src/lib/mono/shape/shape-clipboard-manager.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Flow } from "$libs/flow";
import { Rect } from "$libs/graphics-geo/rect";
import { TextExtra } from "$mono/shape/extra/shape-extra";
import {
    type SerializableLineConnector,
    SerializableLineConnectorArraySerializer,
} from "$mono/shape/serialization/connector";
import { Jsonizable, Serializer, SerialName } from "$mono/shape/serialization/serializable";
import {
    type AbstractSerializableShape,
    SerializableText,
    ShapeArraySerializer,
} from "$mono/shape/serialization/shapes";
import { AbstractShape } from "./shape/abstract-shape";

/**
 * A clipboard manager specializing for shapes.
 * This class handles storing shapes to clipboard and getting shapes in clipboard from paste action.
 */
export class ShapeClipboardManager {
    private clipboardShapeMutableFlow: Flow<ClipboardObject> = new Flow(ClipboardObject.create([], []));
    public clipboardShapeFlow: Flow<ClipboardObject> = this.clipboardShapeMutableFlow.immutable();

    constructor() {
        document.onpaste = (event) => {
            event.preventDefault();
            event.stopPropagation();
            this.onPasteText(event.clipboardData?.getData('text/plain') || '');
        };
    }

    private onPasteText(text: string) {
        if (text.trim() === '') {
            return;
        }
        const json = this.parseJsonFromString(text);
        // @ts-expect-error fromJson is attached by Jsonizable
        const clipboardObject = json ? ClipboardObject.fromJson(json) : null;
        if (clipboardObject && clipboardObject.shapes.length > 0) {
            this.clipboardShapeMutableFlow.value = clipboardObject;
            return;
        }

        this.clipboardShapeMutableFlow.value = ClipboardObject.create([this.createTextShapeFromText(text)], []);
    }

    private parseJsonFromString(text: string) {
        try {
            const json = JSON.parse(text);
            // Backward compatibility with old clipboard format
            return Array.isArray(json) ? { shapes: json, connectors: [] } : json;
        } catch (e) {
            console.error('Failed to parse JSON from clipboard text', e);
            return null;
        }

    }

    private createTextShapeFromText(text: string): SerializableText {
        const lines = text.split('\n').flatMap(line => line.match(/.{1,400}/g) || []);
        const width = Math.max(...lines.map(line => line.length));
        const height = lines.length;

        const toBeUsedText = text.replace(/ /g, '\u00a0');
        return SerializableText.create({
            id: null,
            isIdTemporary: false,
            versionCode: AbstractShape.nextVersionCode(),
            bound: Rect.byLTWH(0, 0, width, height),
            text: toBeUsedText,
            extra: TextExtra.NO_BOUND.toSerializableExtra(),
            isTextEditable: false,
        });
    }

    public setClipboard(clipboardObject: ClipboardObject) {
        // @ts-expect-error toJson is attached by Jsonizable
        const text = JSON.stringify(clipboardObject.toJson());
        this.setClipboardText(text);
    }

    public setClipboardText(text: string) {
        const textArea = document.createElement('textarea');
        textArea.style.position = 'absolute';
        textArea.style.left = '-10000px';
        textArea.style.top = '-10000px';
        textArea.style.width = '1px';
        textArea.style.height = '1px';
        textArea.style.opacity = '0';
        textArea.value = text;
        document.body.appendChild(textArea);
        textArea.select();
        // TODO: Address this deprecated method
        document.execCommand('copy');
        document.body.removeChild(textArea);
    }
}

/**
 * A data class to store shapes and connectors in clipboard when serializing to JSON.
 */
@Jsonizable
export class ClipboardObject {
    @SerialName("shapes")
    @Serializer(ShapeArraySerializer)
    public shapes: AbstractSerializableShape[] = [];
    @SerialName("connectors")
    @Serializer(SerializableLineConnectorArraySerializer)
    public connectors: SerializableLineConnector[] = [];

    private constructor() {
    }

    static create(shapes: AbstractSerializableShape[], connectors: SerializableLineConnector[]): ClipboardObject {
        const result = new ClipboardObject();
        result.shapes = shapes;
        result.connectors = connectors;
        return result;
    }
}
